src/serde/de_br.rs

```diff
@@ -83,8 +135,113 @@ pub fn node_from_bytes_backrefs_record(
+    let mut arg_index: usize = if parsing_sexp { 0 } else { args.len() - 1 };
+
+    // find first non-zero byte
+    let first_bit_byte_index = first_non_zero(node_index);
+    if first_bit_byte_index >= node_index.len() {
+        return Ok(NodePtr::NIL);
+    }
+
+    // find first non-zero bit (the most significant bit is a sentinel)
+    let last_bitmask = msb_mask(node_index[first_bit_byte_index]);
+
+    // follow through the bits, moving left and right
+    let mut byte_idx = node_index.len() - 1;
+    let mut bitmask = 0x01;
+
+    // if we move from parsing the Vec stack to parsing the SExp stack use the following variable
+    let mut sexp_to_parse = NodePtr::NIL;
+
+    while byte_idx > first_bit_byte_index || bitmask < last_bitmask {
+        let is_bit_set: bool = (node_index[byte_idx] & bitmask) != 0;
+        if parsing_sexp {
+            match allocator.sexp(sexp_to_parse) {
+                SExp::Atom => {
+                    return Err(EvalErr(
+                        sexp_to_parse,
+                        "invalid backreference during deserialisation".into(),
+                    )
+                    .into());
+                }
+                SExp::Pair(left, right) => {
+                    sexp_to_parse = if is_bit_set { right } else { left };
+                }
+            }
+        } else if is_bit_set {
+            // we have traversed right ("rest"), so we keep processing the Vec
+            if arg_index == 0 {
+                // if we have reached the end of the stack, we must start parsing as NIL
+                parsing_sexp = true;
+            } else {
+                // pop from the stack
+                arg_index -= 1;
+            }
+        } else {
+            // we have traversed left (i.e "first" rather than "rest") so we must process as SExp now
+            parsing_sexp = true;
+            sexp_to_parse = args[arg_index].0;
+        }
+
+        if bitmask == 0x80 {
+            bitmask = 0x01;
+            byte_idx -= 1;
+        } else {
+            bitmask <<= 1;
+        }
+    }
+
+    if parsing_sexp {
+        return Ok(sexp_to_parse);
+    } else if args[arg_index].1.is_some() {
+        // cached entry exists - use that instead of recreating the pairs
+        return Ok(args[arg_index].1.unwrap());
+    }
+    // take bottom of stack and make (item . NIL)
+    // or fetch from cache if it exists
+    let mut backref_node = match args[0].1 {
+        Some(cached) => cached,
+        None => {
+            let node = allocator.new_pair(args[0].0, NodePtr::NIL)?;
+            allocator.increase_pair_max(1)?;
+            args[0].1 = Some(node);
+            node
+        }
+    };
+
+    if arg_index == 0 {
+        args[arg_index].1 = Some(backref_node);
+        return Ok(backref_node);
+    }
+
+    // for the rest of items starting from last + 1 in stack
+    for x in args.iter_mut().take(arg_index + 1).skip(1) {
```

Comment:
I don't think you need to treat the bottom element of the stack as a special case. If you start with `backref_node = NodePtr::NIL` you can just loop over `0..arg_index` and create pairs and update the cache, if needed.

Reply:
Done